Refactoring Map.cs to Enable Mod-defined Map Formats

AspectInteractive2 opened the pull request:
This is a rebased and working version of Ice Reaper's PR #18378, with the addition of splitting out the functionality of `Map.cs` into optional plug-and-play interfaces. Together this allows the creation of Map formats, which are custom versions of `Map.cs` that may or may not implement all the existing Map.cs functions (depending on whether they are needed by the mod), and additionally can decide on how these functions are implemented. The most notable example of this currently is isometric vs rectangular cells, these can be defined as two different map formats implementing the same interface, removing the need for numerous 'if else' clauses within the existing Map.cs to check for which C&C mod is currently being used. Additionally this also means that most of the existing functionality in Map.cs can now belong to OpenRA.Mods.Common, which enables users of the mod SDK to modify this behaviour without having to fork the OpenRA repo.

Below are the changes:

- A new interface `IMap` now incorporates all the base functionality of `Map.cs`
- All core functions within OpenRA now point to `IMap`, rather than the `Map` class directly
- All height/ramp related functions now point to a cast of Map to `IMapElevation` rather than to `Map` directly
- All Tiles and Resources related functions now point to casts of Map to `IMapTiles` and `IMapResources` respectively
- `DefaultMap` and `DefaultMapLoader` have been created as the initial Map format, and map format loader, however these will eventually be replaced in future PRs with more specialised map formats (e.g. Gen1Map, Gen1MapLoader for TD/RA and Gen2Map, Gen2MapLoader for TS). This implementation sits in OpenRA.Mods.Common rather than OpenRA.Game.

Also please see [this comment](https://github.com/OpenRA/OpenRA/pull/18378#issuecomment-657236223) from pchote for the motivation behind most of the above. The majority of these changes have been implemented in this PR.

NOTE: I have kept my commits unsquashed for now, to make it easier to split out the PR in the future if this becomes necessary.

AspectInteractive2 committed:
Abstracted map loading to DefaultMapLoader : IMapLoader

AspectInteractive2 committed:
Implementation of map format interfaces IMap, IMapTiles, IMapResources, and IMapElevation

PunkPun commented on OpenRA.Game/ModData.cs:
```diff
@@ -161,16 +164,17 @@ public Map PrepareMap(string uid)
 			Map map;
 			using (new Support.PerfTimer("Map"))
-				map = new Map(this, MapCache[uid].Package);
+				map = (Map)MapLoader.Load(this, MapCache[uid].Package);
```
Redundant cast

AspectInteractive2 replied:
Fixed, was a carryover from a prior version where the cast was needed.

PunkPun commented on OpenRA.Mods.Cnc/Traits/ConyardChronoReturn.cs:
```diff
@@ -147,8 +147,8 @@ void TriggerVortex()
-			var max = chronosphere.World.Map.Grid.MaximumTileSearchRange;
-			foreach (var tile in self.World.Map.FindTilesInCircle(destination, max))
+			var max = (chronosphere.World.Map).Grid.MaximumTileSearchRange;
+			foreach (var tile in (self.World.Map).FindTilesInCircle(destination, max))
```
redundant parentheses

AspectInteractive2 replied:
Fixed, same as above.

PunkPun commented on OpenRA.Mods.Cnc/Scripting/Properties/ChronosphereProperties.cs:
```diff
@@ -36,8 +36,7 @@ public void Chronoshift(LuaTable unitLocationPairs, int duration = 0, bool killC
 
-				var cs = actor.TraitsImplementing<Chronoshiftable>()
-					.FirstEnabledConditionalTraitOrDefault();
+				var cs = actor.TraitsImplementing<Chronoshiftable>().FirstEnabledConditionalTraitOrDefault();
```
redundant change

AspectInteractive2 replied:
Reverted, though this line break is unnecessary in the current source (line length is short enough without it).

PunkPun commented on OpenRA.Mods.Common/MapFormats/DefaultMap.cs:
```diff
@@ -0,0 +1,1432 @@
+				if (yamlStream == null)
+					throw new FileNotFoundException("Required file map.yaml not present in this map");
+
+				// yaml = new MiniYaml(null, MiniYaml.FromStream(yamlStream, "map.yaml")).ToDictionary();
```
?

AspectInteractive2 replied:
Removed comment, was a prior version of this code that I forgot to remove.

PunkPun commented on OpenRA.Game/Map/MapPreview.cs:
```diff
@@ -301,106 +302,11 @@ public MapPreview(Map map, ModData modData)
-				using (var dataStream = p.GetStream("map.png"))
-					newData.Preview = new Png(dataStream);
-
-			newData.ModifiedDate = File.GetLastWriteTime(p.Name);
```
This is never added back?

AspectInteractive2 replied:
See line 693 of `DefaultMap.cs`. Though if you mean `newData.ModifiedDate` I will add that in.

AspectInteractive2 replied:
Has now been added in.

PunkPun commented on OpenRA.Game/Map/Map.cs:
```diff
@@ -212,89 +61,20 @@ public class Map : IReadOnlyFileSystem
-			try
-			{
-				foreach (var filename in contents)
-					if (filename.EndsWith(".yaml") || filename.EndsWith(".bin") || filename.EndsWith(".lua") || (format >= 12 && filename == "map.png"))
```
This is never added back

PunkPun replied:
see #19891

AspectInteractive2 replied:
Have now added this back in as well as the other linked changes needed from that PR.

PunkPun commented on OpenRA.Game/Map/MapPreview.cs:
```diff
@@ -273,16 +273,17 @@ public MapPreview(Map map, ModData modData)
 			{
-				MapFormat = map.MapFormat,
+				// TO DO: Check if this MapFormat retrieval is correct and working
+				MapFormat = Map.GetMapFormat(Package),
```
GetMapFormat is only used to get the format before yaml parsing. You are using it after

AspectInteractive2 replied:
Fixed, this now uses a cast of `IMap` as with the other attributes.

PunkPun replied:
Still unresolved. Keep the original implementation

PunkPun commented on OpenRA.Game/Map/MapGrid.cs:
```diff
@@ -123,7 +123,7 @@ public class MapGrid : IGlobalModData
 		public CellRamp[] Ramps { get; }
 
-		internal readonly CVec[][] TilesByDistance;
+		public CVec[][] TilesByDistance { get; private set; }
```
```suggestion
		public readonly CVec[][] TilesByDistance;
```

?

AspectInteractive2 replied:
Both seem to be equivalent but sure, has been updated.

PunkPun commented on OpenRA.Game/Map/MapPreview.cs:
```diff
@@ -273,16 +273,17 @@ public MapPreview(Map map, ModData modData)
 			innerData = new InnerData
 			{
-				MapFormat = map.MapFormat,
+				// TO DO: Check if this MapFormat retrieval is correct and working
```
now this commend is redundant

AspectInteractive2 replied:
Yep, fixed

PunkPun commented on mods/modcontent/mod.yaml:
```diff
@@ -73,3 +73,5 @@ TerrainFormat: DefaultTerrain
 
 ModelSequenceFormat: PlaceholderModelSequence
+
+DefaultMapLoader:
```
Add a newline so it's like all the other files

AspectInteractive2 replied:
Added a new line after `DefaultMapLoader:`, I think that's what you mean?

PunkPun commented on OpenRA.Mods.Common/UtilityCommands/Utilities.cs:
```diff
@@ -32,7 +32,8 @@ public static class Utilities
 				{
-					map = new Map(modData, new Folder(Platform.EngineDir).OpenPackage(mapPath, modData.ModFiles));
+					// TO DO: Make this use IMap instead of Map
+					map = (Map)modData.MapLoader.Load(modData, new Folder(Platform.EngineDir).OpenPackage(mapPath, modData.ModFiles));
```
Redundant cast. And for the comment, are there any plans?

AspectInteractive2 replied:
No plans as of yet but I think it is worth keeping for a future PR potentially.

PunkPun commented on OpenRA.Mods.Common/Lint/CheckMapMetadata.cs:
```diff
@@ -10,26 +10,30 @@
-				emitError($"Map format {mapFormat} does not match the supported version {Map.CurrentMapFormat}.");
+			if (mapFormat != DefaultMap.SupportedMapFormat)
+				emitError("Map format {0} does not match the supported version {1}."
+					.F(mapFormat, DefaultMap.SupportedMapFormat));
```
Revert these changes

AspectInteractive2 replied:
Have brought it back however I am not sure what `CurrentMapFormat` means since there is simply `MapFormat` in the new version which is checked against the `SupportedMapFormat`. May need some clarification on this.

PunkPun commented on OpenRA.Game/Map/Map.cs:
```diff
@@ -65,137 +25,26 @@ public enum MapVisibility
+	public abstract class Map : IReadOnlyFileSystem
 	{
-		public const int SupportedMapFormat = 11;
-		public const int CurrentMapFormat = 12;
```
This is never added back

AspectInteractive2 replied:
Once again not sure what is meant here, `SupportedMapFormat` exists and `MapFormat` is used for the map's format.

PunkPun replied:
See #19891

AspectInteractive2 replied:
Have added this back now.

PunkPun commented on OpenRA.Mods.Common/UtilityCommands/ExtractMapRules.cs:
```diff
@@ -17,10 +17,10 @@ namespace OpenRA.Mods.Common.UtilityCommands
 	public class ExtractMapRules : IUtilityCommand
 	{
-		string IUtilityCommand.Name => "--map-rules";
+		string IUtilityCommand.Name => "--imap-rules";
```
We probably shouldn't utility change commands

AspectInteractive2 replied:
Fixed

PunkPun commented on OpenRA.Mods.Common/UtilityCommands/ExtractMapRules.cs:
```diff
@@ -47,19 +47,20 @@ void MergeAndPrint(Map map, string key, MiniYaml value)
 		}
 
-		[Desc("MAPFILE", "Merge custom map rules into a form suitable for including in map.yaml.")]
+		[Desc("MAPFILE", "Merge custom map rules into a form suitable for including in imap.yaml.")]
```
the file is still called map.yaml

AspectInteractive2 replied:
Fixed

PunkPun commented on OpenRA.Mods.Common/Traits/World/EditorResourceLayer.cs:
```diff
@@ -123,11 +123,14 @@ public void WorldLoaded(World w, WorldRenderer wr)
 		{
-			var uv = cell.ToMPos(Map);
-			if (!Map.Resources.Contains(uv))
+			var map = Map;
```
redundant variable

AspectInteractive2 replied:
Done

PunkPun commented on OpenRA.Mods.Common/Traits/World/EditorCursorLayer.cs:
```diff
@@ -107,7 +107,7 @@ void ITickRender.TickRender(WorldRenderer wr, Actor self)
 						updated = true;
 
-					var subcell = world.Map.Tiles.Contains(cell) ? editorLayer.FreeSubCellAt(cell) : SubCell.Invalid;
+					var subcell = world.Map.Contains(cell) ? editorLayer.FreeSubCellAt(cell) : SubCell.Invalid;
```
```suggestion
					var subcell = ((IMapTiles)world.Map).Tiles.Contains(cell) ? editorLayer.FreeSubCellAt(cell) : SubCell.Invalid;
```

?

AspectInteractive2 replied:
Done, wasn't sure they were equivalent but yes this works.

PunkPun commented on OpenRA.Mods.Common/Lint/CheckMapMetadata.cs:
```diff
@@ -10,26 +10,30 @@
 		{
-			Run(emitError, map.MapFormat, map.Author, map.Title, map.Categories);
+			var map = (Map)imap;
+			var defaultMap = imap as DefaultMap;
```
just inline this variable

AspectInteractive2 replied:
Done

PunkPun commented on OpenRA.Mods.Common/EditorBrushes/EditorCopyPasteBrush.cs:
```diff
@@ -201,9 +203,10 @@ class CopyPasteEditorAction : IEditorAction
-			mapTiles = map.Tiles;
-			mapHeight = map.Height;
-			mapResources = map.Resources;
+			var mapResource = (IMapResource)map;
```
redundant variable

AspectInteractive2 replied:
Done

PunkPun commented on OpenRA.Game/World.cs:
```diff
@@ -208,7 +208,7 @@ internal World(ModData modData, Map map, OrderManager orderManager, WorldType ty
 			LocalRandom = new MersenneTwister();
 
-			ModelCache = modData.ModelSequenceLoader.CacheModels(map, modData, map.Rules.ModelSequences);
+			ModelCache = modData.ModelSequenceLoader.CacheModels(map, modData, ((IMap)map).Rules.ModelSequences);
```
```suggestion
			ModelCache = modData.ModelSequenceLoader.CacheModels(map, modData, Map.Rules.ModelSequences);
```

AspectInteractive2 replied:
Done

PunkPun commented on OpenRA.Game/World.cs:
```diff
@@ -233,11 +233,11 @@ internal World(ModData modData, Map map, OrderManager orderManager, WorldType ty
-				MapUid = Map.Uid,
-				MapTitle = Map.Title
+				MapUid = ((Map)Map).Uid,
+				MapTitle = ((Map)Map).Title
```
And here

AspectInteractive2 replied:
Done

PunkPun commented on OpenRA.Game/World.cs:
```diff
@@ -233,11 +233,11 @@ internal World(ModData modData, Map map, OrderManager orderManager, WorldType ty
 			};
 
-			RulesContainTemporaryBlocker = map.Rules.Actors.Any(a => a.Value.HasTraitInfo<ITemporaryBlockerInfo>());
+			RulesContainTemporaryBlocker = ((IMap)map).Rules.Actors.Any(a => a.Value.HasTraitInfo<ITemporaryBlockerInfo>());
```
Same here

AspectInteractive2 replied:
Done

PunkPun commented on OpenRA.Game/ModData.cs:
```diff
@@ -161,16 +164,17 @@ public Map PrepareMap(string uid)
 			if (MapCache[uid].Status != MapStatus.Available)
 				throw new InvalidDataException($"Invalid map uid: {uid}");
 
+			// TO DO: Make loader use IMap interface instead of Map
```
```suggestion
			// TODO: Make loader use IMap interface instead of Map
```
This is how TODO is written everywhere in the engine. Please change the 4 comments to this

AspectInteractive2 replied:
Fixed

PunkPun reviewed (commented):
Creating a new map crashes

AspectInteractive2 commented:
> Creating a new map crashes

This was due to a typo in the If statement for Grid.MaximumTerrainHeight > 0 in `DefaultMap.cs`, as well as `MapSize` not being initialised. It now works.

anvilvapre commented on OpenRA.Game/Graphics/Viewport.cs:
```diff
@@ -294,7 +295,7 @@ public CPos ViewToWorld(int2 view)
 
 			// Something is very wrong, but lets return something that isn't completely bogus and hope the caller can recover
-			return worldRenderer.World.Map.CellContaining(worldRenderer.ProjectedPosition(ViewToWorldPx(view)));
+			return (worldRenderer.World.Map).CellContaining(worldRenderer.ProjectedPosition(ViewToWorldPx(view)));
```
unneeded change?

AspectInteractive2 replied:
Fixed

anvilvapre commented on OpenRA.Game/Graphics/WorldRenderer.cs:
```diff
@@ -48,8 +48,9 @@ public sealed class WorldRenderer : IDisposable
-			TileScale = World.Map.Grid.Type == MapGridType.RectangularIsometric ? 1448 : 1024;
+			var map = World.Map;
+			TileSize = map.Grid.TileSize;
+			TileScale = map.Grid.Type == MapGridType.RectangularIsometric ? 1448 : 1024;
```
suggestion to then also hide this in one of the IMap interfaces. seems wrong to hardcode these constants in world renderer.

AspectInteractive2 replied:
These are actually within `IMap` now, as the WorldRenderer now uses an `IMap` for its map variable.

![image](https://user-images.githubusercontent.com/1495872/209916248-a15f9841-67b8-4431-a99b-9974bfd6b0c2.png)

anvilvapre commented on OpenRA.Game/Map/IMap.cs:
```diff
@@ -0,0 +1,76 @@
+		byte[] SavePreview();
+	}
+
+	public interface IMapElevation : IMap
```
code above does not check if a map implements this interface. does a direct cast. likely because also maps without elevation implement it(?).

may as well add this to IMap.

AspectInteractive2 replied:
Not sure if you referenced the right code above as I don't see the cast. However the idea is that functions that are not implemented would not be called in the first place, because the mod would not call them. E.g.: Elevation isn't used in Red Alert, so it would never call functions that have casts to `IMapElevation`.

anvilvapre commented on OpenRA.Game/Map/MapPreview.cs:
```diff
@@ -273,16 +273,16 @@ public MapPreview(Map map, ModData modData)
 			innerData = new InnerData
 			{
-				MapFormat = map.MapFormat,
+				MapFormat = ((IMap)map).MapFormat,
```
unnessary cast? i assume map implements IMap.

AspectInteractive2 replied:
Map does not implement `IMap`, `DefaultMap` does. `Map` now only has the core functionality central to all maps, which is basically loading and checking the map, and getting its map format (see the new Map.cs for the remaining functionality of `Map`, which is now only 164 lines of code.

PunkPun replied:
Rebase fail. The first commit makes a change that is then unmade

AspectInteractive2 replied:
This is correct because the `IMap` interface is first introduced in the second commit. First commit only deals with abstracting the map loader.

anvilvapre commented on OpenRA.Game/Map/MapPreview.cs:
```diff
@@ -273,16 +273,16 @@ public MapPreview(Map map, ModData modData)
 				Categories = map.Categories,
 				Author = map.Author,
-				TileSet = map.Tileset,
+				TileSet = ((IMap)map).Tileset,
```
Same? Also 2 times more below.

AspectInteractive2 replied:
See above

anvilvapre commented on OpenRA.Game/Player.cs:
```diff
@@ -156,7 +156,7 @@ public Player(World world, Session.Client client, PlayerReference pr, MersenneTw
 			PlayerReference = pr;
 
-			inMissionMap = world.Map.Visibility.HasFlag(MapVisibility.MissionSelector);
+			inMissionMap = ((Map)world.Map).Visibility.HasFlag(MapVisibility.MissionSelector);
```
no interface for this? IMissionSelector, Manifest.GetMissionSelector?

AspectInteractive2 replied:
Perhaps could be separated out but this is simply a flag for which map lists the map should appear on. This will be easy enough to do after the fact in a later PR once the main refactoring is complete. I think it is not needed for this PR which is already very large.

anvilvapre commented on OpenRA.Game/Scripting/ScriptContext.cs:
```diff
@@ -201,7 +201,7 @@ public sealed class ScriptContext : IDisposable
 			{
 				foreach (var s in scripts)
-					loadScript.Call(s, world.Map.Open(s).ReadAllText()).Dispose();
+					loadScript.Call(s, ((Map)world.Map).Open(s).ReadAllText()).Dispose();
```
if its not high performance code, i would try a cast. this way future mods can keep using default functionality.

AspectInteractive2 replied:
Don't understand, the above is casting to `Map`. If you mean cast to `IMap`, this would be difficult as I would have to change the Lua handling part of the code base. I don't think this is needed for the purpose of this PR, which is primarily to introduce map formats and abstract away some main functions of Maps into interfaces. Once again it could be looked at in a future PR.

anvilvapre commented on OpenRA.Game/Support/ExceptionHandler.cs:
```diff
@@ -38,7 +38,7 @@ public static void HandleFatalError(Exception ex)
 			{
 				var map = Game.OrderManager.World.Map;
-				Log.Write("exception", $"on map {map.Uid} ({map.Title} by {map.Author}).");
+				Log.Write("exception", $"on map {((Map)map).Uid} ({((Map)map).Title} by {((Map)map).Author}).");
```
also feels like a missing map interface i.e. IMapInfo { Uid, Tititle, Author }, IMap.GetMapInfo() or GetMapAttributes().

optionally return them in a dictionary as key value pairs.

AspectInteractive2 replied:
The purpose of interfaces is to either A) Provide alternative implementations, or B) Provide the option to not implement the interface at all and not use the subsequent functionality. Since all maps have Titles and Authors, and all maps will implement a Title and Author in the same way, I don't see the value in adding an interface here. Perhaps in a later PR if there is a case for it, but I don't think this is critical to the current refactoring plan.

PunkPun replied:
casts could be moved higher up

anvilvapre commented on OpenRA.Game/Traits/Player/Shroud.cs:
```diff
@@ -141,7 +141,7 @@ public bool Disabled
 		{
 			this.info = info;
-			map = self.World.Map;
+			map = (IMap)self.World.Map;
```
unnessecarry?

AspectInteractive2 replied:
Fixed

anvilvapre commented on OpenRA.Game/Traits/Target.cs:
```diff
@@ -46,7 +46,7 @@ public enum TargetType : byte { Invalid, Actor, Terrain, FrozenActor }
 		{
 			type = TargetType.Terrain;
-			terrainCenterPosition = w.Map.CenterOfSubCell(c, subCell);
+			terrainCenterPosition = ((IMap)w.Map).CenterOfSubCell(c, subCell);
```
unnessary?

AspectInteractive2 replied:
Fixed

anvilvapre commented on OpenRA.Game/Traits/World/ScreenMap.cs:
```diff
@@ -63,7 +63,7 @@ public class ScreenMap : IWorldLoaded
 		public ScreenMap(World world, ScreenMapInfo info)
 		{
-			var size = world.Map.Grid.TileSize;
+			var size = ((IMap)world.Map).Grid.TileSize;
```
same?

AspectInteractive2 replied:
Fixed

anvilvapre commented on OpenRA.Game/World.cs:
```diff
@@ -189,7 +189,7 @@ internal World(ModData modData, Map map, OrderManager orderManager, WorldType ty
 			Type = type;
 			OrderManager = orderManager;
-			Map = map;
+			Map = (IMap)map;
```
Why not pass an IMap in constructor.

AspectInteractive2 replied:
There are two reasons. First `World.cs` is part of `OpenRA.Game`, so it is not mod code and therefore may need to use `Map.cs` without the interface. Secondly it does call some functions that rely on `Map` not `IMap`, such as `CacheModels`, which is out of scope to refactor at this stage. However since `World.cs` will likely always need to be in `OpenRA.Game`, there will always be core map functions that it can and will need to access that are not part of `IMap`, so a cast to `IMap` while retaining the `Map` parameter makes the most sense here.

anvilvapre commented on OpenRA.Mods.Cnc/Traits/Render/WithBuildingBib.cs:
```diff
@@ -56,7 +56,7 @@ public IEnumerable<IActorPreview> RenderPreviewSprites(ActorPreviewInitializer i
 				// Some mods may define terrain-specific bibs
 				var sequence = Sequence;
-				if (map.Tiles.Contains(cell))
+				if (((IMapTiles)map).Tiles.Contains(cell))
```
to me it seems more logical to let a map own a TileManager. i.e. map.GetTileManager(). promoting splitting logic.

i.e. otherwise you could have a map implement endless interfaces.

AspectInteractive2 replied:
Map no longer has `Tiles`, this functionality now belongs to `IMapTiles` and the current implementation sits in `DefaultMap` rather than `Map`. And `Tiles` is a CellLayer class which seems to work the same as your proposed `TileManager` class. In any case changing how Tiles are managed is not the goal of this PR. All I am doing here is moving the functionality out of `Map.cs` and into `DefaultMap`, behind an interface `IMapTiles`.

anvilvapre replied:
i understand. but then still, it does not make sense to have IMapTiles inherit from IMap. Completely different things. Like letting ITrait inherit from IWidget. just because some class will implement both.

but i understand how you came to this solution.

AspectInteractive2 replied:
I agree, but this is how it was done with the Tilesets refactoring as well (see below link), with `ITemplatedTerrainInfo` and `ITerrainInfoNotifyMapCreated` both inheriting from `ITerrainInfo`. The downside is it adds a dependency to `IMap` where it may not need to be there, but on the plus side it means less casting is necessary, since you can cast to `IMapTiles` and still use the functionality from `IMap`, rather than having to cast twice.

https://github.com/OpenRA/OpenRA/pull/18728/files#diff-b08312ffb1592016a1b6797006dca462f3eb0f22a38ca2956e55e1e9c4f49ee1R14-R23

anvilvapre commented on OpenRA.Mods.Cnc/Traits/World/TSEditorResourceLayer.cs:
```diff
@@ -42,6 +42,8 @@ public TSEditorResourceLayer(Actor self, TSEditorResourceLayerInfo info)
 
 		bool IsValidVeinNeighbour(CPos cell, CPos neighbour)
 		{
+			var mapRamp = ((IMapElevation)Map).Ramp;
```
try the cast, to allows mods to keep using this logic, even if disabled.

AspectInteractive2 replied:
Not sure what you mean here. Mods can keep using this logic if they want and it is still part of the default functionality found in `DefaultMap`.

pchote replied:
Likewise here.

anvilvapre commented on OpenRA.Mods.Cnc/Traits/World/TSEditorResourceLayer.cs:
```diff
@@ -50,30 +52,30 @@ bool IsValidVeinNeighbour(CPos cell, CPos neighbour)
 		protected override bool AllowResourceAt(string resourceType, CPos cell)
 		{
-			var mapResources = Map.Resources;
+			var mapResources = ((IMapResource)Map).Resources;
```
`map.GetResourceManager()?.Resources`

AspectInteractive2 replied:
See my comment above about `IMapTiles`, this functionality is also no longer in `Map`, and we want this functionality to be behind the interface `IMapResource` so that map formats have the choice on whether to implement it or not.

anvilvapre reviewed (commented):
Change makes sense.

Don't find it a logical choice to keep map implementing all the logic of all interfaces. Better would be to isolate the logic in other objects i.e. Map.GetTileManager().Tiles. But i understand that makes the rewrite even bigger. 

Casts are now done without checks. Because the default mods are known to implement them. But to support mods, better would be to try the cast and skip the functionality if not implemented by the mod.

PunkPun commented on OpenRA.Mods.Common/Widgets/Logic/Editor/NewMapLogic.cs:
```diff
@@ -57,16 +57,17 @@ public NewMapLogic(Action onExit, Action<string> onSelect, Widget widget, World
 				var maxTerrainHeight = world.Map.Grid.MaximumTerrainHeight;
 				var tileset = modData.DefaultTerrainInfo[tilesetDropDown.Text];
-				var map = new Map(Game.ModData, tileset, width + 2, height + maxTerrainHeight + 2);
+				var map = Game.ModData.MapLoader.Create(Game.ModData, tileset, width + 2, height + maxTerrainHeight + 2);
```
```suggestion
				var map = modData.MapLoader.Create(modData, tileset, width + 2, height + maxTerrainHeight + 2);
```

?

AspectInteractive2 replied:
Sure, your suggested method is better.

PunkPun commented on OpenRA.Mods.Common/UtilityCommands/ExtractMapRules.cs:
```diff
@@ -52,14 +52,15 @@ void IUtilityCommand.Run(Utility utility, string[] args)
-			MergeAndPrint(map, "Voices", map.VoiceDefinitions);
-			MergeAndPrint(map, "Music", map.MusicDefinitions);
-			MergeAndPrint(map, "Notifications", map.NotificationDefinitions);
+			var map = modData.MapLoader.Load(modData, new Folder(".").OpenPackage(args[1], modData.ModFiles));
```
```suggestion
			var map = modData.MapLoader.Load(modData, new Folder(Platform.EngineDir).OpenPackage(args[1], modData.ModFiles));
```

Shouldn't it be?

AspectInteractive2 replied:
Yes, fixed.

PunkPun replied:
fixed in the wrong commit

AspectInteractive2 replied:
Fixed in the correct commit now.

PunkPun commented on OpenRA.Mods.Common/MapFormats/DefaultMap.cs:
```diff
@@ -0,0 +1,1451 @@
+			try
+			{
+				foreach (var filename in contents)
+					if (filename.EndsWith(".yaml") || filename.EndsWith(".bin") || filename.EndsWith(".lua") || (format >= 12 && filename == "map.png"))
```
Rebase fail. The first commit doesn't have the `(format >= 12 && filename == "map.png")`

AspectInteractive2 replied:
> Reviewed the first commit. Running `make` throws tons of errors, and the commit should be renamed to something like: "Abstracted map loading to DefaultMapLoader : IMapLoader"

Yes there have been numerous PRs since then which prevents this change from working in isolation. Also because of how tightly coupled the loader is with the new map format, I would have to revert the new map format changes to have this commit work in isolation. However the changes that were made are still valid and can be reviewed in isolation, provided that you still refer to the latest commit for the working version.

PunkPun replied:
There are more rebase issues surrounding formats

PunkPun commented on OpenRA.Game/Map/Map.cs:
```diff
@@ -317,1050 +97,14 @@ static int GetMapFormat(IReadOnlyPackage p)
 			// Will be dropped on save if nothing is added to it
 			RuleDefinitions = new MiniYaml("");
-
-			Tiles = new CellLayer<TerrainTile>(Grid.Type, size);
```
Rebase fail. First commit comments this out

PunkPun reviewed (commented):
Reviewed the first commit. Running `make` throws tons of errors, and the commit should be renamed to something like: "Abstracted map loading to DefaultMapLoader : IMapLoader"

PunkPun commented:
It's weird how you added a massive interface of IMapCell.cs, even adding a new file in third commit and then removed it in 6th commit. There are a bunch of errors in the first commits that then get fixed in following commits, even when they couldn't have not been introduced in the first place

AspectInteractive2 commented:
> It's weird how you added a massive interface of IMapCell.cs, even adding a new file in third commit and then removed it in 6th commit. There are a bunch of errors in the first commits that then get fixed in following commits, even when they couldn't have not been introduced in the first place

The purpose of keeping the commit is purely to show the genesis of the project, since it went through a revision of separating out the `IMapCell` interface, before merging it back into `IMap` after discussing it and finding that the cell logic is too closely coupled to the engine to be isolated at this stage. The first and last commits are the most meaningful, since the first shows the initial loader and map format design, while the latter shows the new interface that is being implemented to support the new map format. All of the commits in between are purely revisions to show the evolution, in case you are wondering how the various interfaces came to be.

PunkPun commented:
How this PR came to be or it's evolution is irrelevant and just noise making it harder to find regressions

AspectInteractive2 commented:
> How this PR came to be or it's evolution is irrelevant and just noise making it harder to find regressions

Sure, okay, I've squashed the commits and renamed the first commit.

PunkPun commented on OpenRA.Mods.Cnc/UtilityCommands/ImportTSMapCommand.cs:
```diff
@@ -383,7 +386,7 @@ static void ReadOverlay(Map map, IniFile file, int2 fullSize)
 			UnpackLCW(overlayDataCompressed, overlayDataPack, temp);
 
-			var overlayIndex = new CellLayer<int>(map);
+			var overlayIndex = new CellLayer<int>((IMap)map);
```
redundant cast

AspectInteractive2 replied:
Fixed

PunkPun commented on OpenRA.Mods.Common/EditorBrushes/EditorCopyPasteBrush.cs:
```diff
@@ -109,10 +109,12 @@ public bool HandleMouseInput(MouseInput mi)
-			var mapResources = worldRenderer.World.Map.Resources;
+			var map = worldRenderer.World.Map;
+			var gridType = map.Grid.Type;
+			var mapResource = (IMapResource)worldRenderer.World.Map;
```
```suggestion
			var mapResource = (IMapResource)map;
```

PunkPun commented on OpenRA.Mods.Common/MapFormats/DefaultMap.cs:
```diff
@@ -0,0 +1,1451 @@
+			Height = new CellLayer<byte>(Grid.Type, size);
+			Ramp = new CellLayer<byte>(Grid.Type, size);
+
+			using (var s = Package.GetStream("map.bin"))
```
this gets commended out in the first commit and then uncommented in the second

AspectInteractive2 replied:
No longer the case

PunkPun commented on OpenRA.Mods.Common/Traits/World/ElevatedBridgeLayer.cs:
```diff
@@ -44,6 +44,8 @@ public ElevatedBridgeLayer(Actor self, ElevatedBridgeLayerInfo info)
 
 		public void WorldLoaded(World world, WorldRenderer wr)
 		{
+			var mapCell = map;
```
this variable isn't needed

AspectInteractive2 replied:
Fixed

PunkPun commented on OpenRA.Mods.Common/Traits/World/LegacyBridgeLayer.cs:
```diff
@@ -67,12 +69,15 @@ public void WorldLoaded(World w, WorldRenderer wr)
 
 		void ConvertBridgeToActor(World w, CPos cell)
 		{
+			var map = w.Map;
```
this variable isn't needed

PunkPun commented on mods/modcontent/mod.yaml:
```diff
@@ -73,3 +73,5 @@ TerrainFormat: DefaultTerrain
 
 ModelSequenceFormat: PlaceholderModelSequence
+
+DefaultMapLoader:
```
should be added in the first commit

AspectInteractive2 replied:
Fixed both of these

PunkPun commented on OpenRA.Game/Map/Map.cs:
```diff
@@ -65,126 +32,18 @@ public enum MapVisibility
+		Map Load(ModData modData, IReadOnlyPackage package);
+		Map Create(ModData modData, ITerrainInfo terrainInfo, int width, int height);
+		string ComputeUID(ModData modData, IReadOnlyPackage package);
+		void UpdatePreview(ModData modData, MapPreview mapPreview, IReadOnlyPackage p, IReadOnlyPackage parent, MapClassification classification, string[] mapCompatibility, MapGridType gridType);
```
couldn't we just pass ModData in the IMapLoader constructor and avoid passing it in every function? MapPreview also contains a reference to MapData so it's unnecessary to pass

PunkPun commented on OpenRA.Game/Map/Map.cs:
```diff
@@ -212,91 +61,20 @@ public class Map : IReadOnlyFileSystem, IDisposable
+#pragma warning disable IDE1006 // Naming Styles
+		protected readonly ModData modData;
+		protected Translation Translation;
+#pragma warning restore IDE1006 // Naming Styles
```
Just follow the rule. There are only a few uses of these variables anyway

AspectInteractive2 replied:
From memory following this rule actually isn't possible because `modData` must be distinct from `ModData`, yet the naming rule does not allow lowercase 'm'. So the only way to follow the rule is to do something like `ModDataInstance`, which would be nastier than disabling the rule for those two lines imo.

PunkPun commented on OpenRA.Game/Map/IMap.cs:
Perhaps rename file to MapInterfaces and move IMapLoader here

PunkPun commented on OpenRA.Game/Map/Map.cs:
```diff
@@ -212,91 +61,20 @@ public class Map : IReadOnlyFileSystem, IDisposable
-		}
-
-		static int GetMapFormat(IReadOnlyPackage p)
+		public static int GetMapFormat(IReadOnlyPackage p)
```
It's weird to have this defined in Map.cs when it is only used in DefaultMap for uid extraction

PunkPun commented on OpenRA.Game/Map/MapPreview.cs:
```diff
@@ -63,8 +63,8 @@ public class RemoteMapData
 	{
-		/// <summary>Wrapper that enables map data to be replaced in an atomic fashion.</summary>
-		class InnerData
+		/// <summary>Wrapper that enables map data to be replaced in an atomic fashion</summary>
```
?

PunkPun commented on OpenRA.Mods.Common/MapFormats/DefaultMap.cs:
```diff
@@ -0,0 +1,536 @@
+			new MapField("Voices", "VoiceDefinitions", required: false),
+			new MapField("Music", "MusicDefinitions", required: false),
+			new MapField("Notifications", "NotificationDefinitions", required: false),
+			//new MapField("Translations", "TranslationDefinitions", required: false)
```
Rebase error

PunkPun commented on OpenRA.Mods.Common/MapFormats/DefaultMap.cs:
```diff
@@ -0,0 +1,536 @@
+
+			if (yaml.TryGetValue("MapFormat", out var temp))
+			{
+				var format = FieldLoader.GetValue<int>("MapFormat", temp.Value);
```
newData.MapFormat is never set

PunkPun commented on OpenRA.Mods.Common/UtilityCommands/CheckYaml.cs:
```diff
@@ -97,7 +97,7 @@ void IUtilityCommand.Run(Utility utility, string[] args)
 						continue;
 
-					using (var testMap = new Map(modData, package))
+					var testMap = modData.MapLoader.Load(modData, package);
```
Keep the `using`

PunkPun commented on OpenRA.Game/Map/Map.cs:
```diff
@@ -25,37 +23,6 @@
-				throw new InvalidDataException($"Unknown binary map format '{Format}'");
-		}
-	}
 
```
"An opening brace should not be followed by a blank line"

penev92 commented on OpenRA.Game/Map/IMap.cs:
```diff
@@ -0,0 +1,78 @@
+		SequenceSet Sequences { get; }
+		CellRegion AllCells { get; }
+		List<CPos> AllEdgeCells { get; }
+		WDist CellHeightStep { get; }
```
Why isn't this on the Elevation interface?

PunkPun replied:
https://github.com/OpenRA/OpenRA/pull/20582#discussion_r1159809521

PunkPun commented on OpenRA.Mods.Common/Lint/CheckMapMetadata.cs:
```diff
@@ -11,25 +11,28 @@
 
 using System;
 using OpenRA.Server;
+using OpenRA.Mods.Common.MapFormats;
```
"Using directives should be ordered alphabetically by the namespaces"

PunkPun commented on OpenRA.Mods.Common/MapFormats/DefaultMap.cs:
```diff
@@ -0,0 +1,536 @@
+	{
+		public const int SupportedMapFormat = 12;
+
+		/// <summary>Defines the order of the fields in map.yaml</summary>
```
"Documentation text should end with a period"

PunkPun commented on OpenRA.Mods.Common/MapFormats/DefaultMap.cs:
```diff
@@ -0,0 +1,536 @@
+			catch (Exception e)
+			{
+				Log.Write("debug", "Failed to load rules for {0} with error {1}", Title, e);
+				rules = Ruleset.LoadDefaultsForTileSet(modData, Tileset);
```
"Unnecessary assignment of a value to 'rules'"

Something went really wrong here. `rules` is set but not used

penev92 commented on OpenRA.Mods.Common/Traits/World/EditorResourceLayer.cs:
```diff
@@ -106,7 +106,7 @@ public EditorResourceLayer(Actor self, EditorResourceLayerInfo info)
 				kv => kv.Key);
 
-			Map.Resources.CellEntryChanged += UpdateCell;
+			((IMapResource)Map).Resources.CellEntryChanged += UpdateCell;
```
A few lines above don't do `Map = self.World.Map;`. You have no idea whether that would give you a map implementation you can use. Check the type and only use if it implements `IMapResource`. Otherwise you are free to throw here so the modder knows not to use this trait with an incompatible map format.
Then remove all the casts like this one.
Repeat that for every affected trait in this commit.

PunkPun commented on OpenRA.Mods.Common/MapFormats/DefaultMap.cs:
```diff
@@ -0,0 +1,1450 @@
+			return DefaultMap.ComputeUID(package);
+		}
+
+		public void UpdatePreview(ModData modData, MapPreview mp, IReadOnlyPackage p, IReadOnlyPackage parent, MapClassification classification, string[] mapCompatibility, MapGridType gridType)
```
This is a terrible name. This doesn't update, but rather sets Map values from MapPreview

PunkPun commented on OpenRA.Game/Map/MapCache.cs:
```diff
@@ -143,8 +143,8 @@ public void LoadMap(string map, IReadOnlyPackage package, MapClassification clas
-						var uid = Map.ComputeUID(mapPackage);
-						previews[uid].UpdateFromMap(mapPackage, package, classification, modData.Manifest.MapCompatibility, mapGrid.Type);
+						var uid = modData.MapLoader.ComputeUID(modData, mapPackage);
+						modData.MapLoader.UpdatePreview(modData, previews[uid], mapPackage, package, classification, modData.Manifest.MapCompatibility, mapGrid.Type);
```
what happened here???

PunkPun commented:
The extraction of `UpdateFromMap` from `MapPreview`  to `DefaultMap` doesn't make any sense. If you wanted customisability you'd need to extract way way more. Atm it's just a redundant change. Ping @IceReaper

PunkPun commented:
I'd say MapPreview is completely out of scope of this PR

pchote commented on OpenRA.Game/Graphics/TerrainSpriteLayer.cs:
```diff
@@ -32,7 +32,7 @@ public sealed class TerrainSpriteLayer : IDisposable
 
 		readonly WorldRenderer worldRenderer;
-		readonly Map map;
+		readonly IMap map;
```
The `TerrainSpriteLayer` constructor should be checking that `world.Map` implements `IMapTiles` and throw an exception if it doesn't (see how the tileset code does this for examples).

pchote commented on OpenRA.Game/Graphics/TerrainSpriteLayer.cs:
```diff
@@ -88,7 +88,7 @@ public void Update(CPos cell, Sprite sprite, PaletteReference palette, float sca
 			if (sprite != null)
 			{
-				var cellOrigin = map.CenterOfCell(cell) - new WVec(0, 0, map.Grid.Ramps[map.Ramp[cell]].CenterHeightOffset);
+				var cellOrigin = map.CenterOfCell(cell) - new WVec(0, 0, map.Grid.Ramps[((IMapElevation)map).Ramp[cell]].CenterHeightOffset);
```
```csharp
var cellOrigin = map.CenterOfCell(cell);
if (map is IMapElevation mapElevation)
    cellOrigin -= new WVec(0, 0, map.Grid.Ramps[mapElevation.Ramp[cell]].CenterHeightOffset);
```

pchote commented on OpenRA.Game/Graphics/Viewport.cs:
```diff
@@ -242,6 +242,7 @@ public CPos ViewToWorld(int2 view)
 		{
 			var world = worldRenderer.Viewport.ViewToWorldPx(view);
 			var map = worldRenderer.World.Map;
+			var mapRamp = ((IMapElevation)map).Ramp;
```
Likewise here, `Viewport` should be able to handle maps that don't implement `IMapElevation` (this is one of the main usecases in the upstream mods for these interfaces!)

pchote commented on OpenRA.Game/Map/MapGrid.cs:
```diff
@@ -126,7 +126,7 @@ public class MapGrid : IGlobalModData
 		public CellRamp[] Ramps { get; }
 
-		internal readonly CVec[][] TilesByDistance;
+		public readonly CVec[][] TilesByDistance;
```
This is only used within `DefaultMap`. Move it there instead of making it public.

pchote commented on OpenRA.Game/Map/ProjectedCellRegion.cs:
```diff
@@ -45,7 +45,7 @@ public ProjectedCellRegion(Map map, PPos topLeft, PPos bottomRight)
 
 			// Use the map Height data array to clamp the bottom coordinate so it doesn't overflow the map
-			mapBottomRight = map.Height.Clamp(new MPos(bottomRight.U, bottomRight.V + heightOffset));
+			mapBottomRight = ((IMapElevation)map).Height.Clamp(new MPos(bottomRight.U, bottomRight.V + heightOffset));
```
This should support maps that don't implement `IMapElevation`.

pchote commented on OpenRA.Mods.Cnc/Traits/World/JumpjetActorLayer.cs:
```diff
@@ -56,7 +56,7 @@ public JumpjetActorLayer(Actor self, JumpjetActorLayerInfo info)
 
 						neighbourCount++;
-						neighbourHeight += map.Height[neighbour];
+						neighbourHeight += ((IMapElevation)map).Height[neighbour];
```
Check that map implements the interface first and throw if it doesn't (see tileset code).

pchote reviewed (commented):
In general (and i'm pretty sure I have already explained this more than once!) this is very difficult to review and has a high risk of regressions because it is attempting to do things backwards and touching a lot of things twice.

Like I showed with the tileset generalisation, the simpler approach is to start by introducing the interfaces, changing the way that `world.Map` is accessed without changing anything about how that object is created. This separates out concerns allowing each interface to be done with its own commit.

Only at the end do we try to replace `Map` with `DefaultMap` etc - none of the users of `world.Map` need to change at this point because they are all already using the interfaces.

A very partial specific review below.

abcdefg30 commented:
Sorry, but I have to agree with @pchote here. I don't think this PR can be merged in it's current state, but should be split into more digestible parts.

PunkPun commented:
Closing as this PR has become too outdated
